### ISSUE TITLE
The allocator's cost cannot rank rungs of one format at matched bytes: Tessera menu ships 2.00x worse than uniform

### ISSUE THREAD

**Issue by RobTand** [RobTand/prismaquant#117]
Measured and **served** 2026-09-02. A PrismaQuant-allocated Tessera checkpoint at 4.0 bpp is **2.00× worse in served KL than a byte-matched uniform arm** at the same bytes. The allocation loop itself is exact — 196/196 bytes reproduce, re-encode is byte-identical — so this is not a plumbing bug: the **cost model cannot rank the rungs of a single format against each other at matched bytes**, and the DP therefore spends its budget in the wrong places.

Consequence: the continuous Tessera menu **cannot ship surrogate-selected**. Either selection runs on real held-out KL (`SELECTION_MODE=validated-surrogate`), or the surrogate learns to rank same-format rungs.

## Why the existing cost does not do it
AURA's `predicted_dloss` is calibrated across *formats* with very different error structure. Tessera's rungs are one format at many rates, where the differences are small, smooth and dominated by rate-allocation effects the unary cost does not see. The screens that rank formats well are exactly the ones with no resolution here.

## Related, and possibly the same root
`FormatSpec`'s byte accounting **special-cased Tessera and overcharged it by 6.25%**, live on both the DP and the byte gate. A format whose price is wrong cannot be ranked correctly against anything, including itself at another rate. Confirm this is fixed on `main` before treating the ranking failure as purely a cost-model question.

## Standing gate this implies
Every rate-axis candidate should be measured beside a **byte-matched uniform arm** — a candidate that cannot beat "spend the same bytes everywhere" has not earned its allocation. Tracked on the Tessera side as RobTand/tessera#3 (standing gate) and RobTand/tessera#1/#2/#4.

## Acceptance
- A cost that reproduces the served KL *ordering* of at least three Tessera rungs at matched bytes on one model, held out.
- The byte-matched uniform arm beside every published rate-axis result.

**Comment by RobTand** [comment]
## Batch `cost-ranking` — the three halves, separately

### 1. "Confirm `FormatSpec` is fixed on `main` before treating this as purely a cost-model question" — confirmed, and it is not the story

Checked in code on `main` (`dfad8d1`), not assumed:

* `FormatSpec.exact_bits_per_param` **replaces** the weight/scale terms rather than adding
  to them — `memory_bytes_for_shape` returns `ceil(n_params · rate / 8)` and never reaches
  the group-scale line (`prismaquant/format_registry.py:245`, `:261`).
* It has since been superseded for Tessera by `bits_for_shape_fn` (`:224`, consumed at
  `:239`/`:253`), because a CHANNEL plane's fp16-per-row and a window body's `2^L`-byte
  table are **per unit**, so no scalar rate can state them and a spec that sets the
  function has no shape-free rate at all.
* Pinned:

  ```
  $ python -m pytest tests/test_tessera_footprint.py tests/test_tessera_shape_dependent_recipe.py -q
  57 passed, 15 warnings in 8.78s
  ```

  including `test_an_exact_rate_is_the_whole_rate_not_a_body_rate` and
  `test_the_closed_form_prices_every_recipe_the_calculator_prices`, which pins the closed
  form against `tessera.calculator.terminal_rate` across the whole grammar.

Two independent confirmations at the artifact, so this does not rest on the tests alone:

* the served run's own check — **196/196 units**, charged bits equal wire bits,
  `1761722368 bits = 4.000260417 bpp` on both sides;
* a **third** accountant, added in this batch: `tessera.control.unit_wire_bits` prices a
  unit through the wire's layout rather than the closed form, and
  `test_the_control_prices_a_unit_exactly_as_prismaquant_charges_for_it` pins it against
  `artifact_bpp` over three families × three shapes.

**So the DP was ranking correctly-priced things.** The 6.25% overcharge is not the root
cause and the ranking failure is a Δloss problem, not a bytes problem.

### 2. "Every rate-axis candidate should be measured beside a byte-matched uniform arm" — built

RobTand/tessera#3, two commits on a Tessera worktree branch (not pushed): `tessera.control`
(library) + `experiments/uniform_control.py` (`plan` / `verify`), with
`plan_from_layer_config.py` pricing the control into every plan sidecar.

The match is **asserted**, not eyeballed: `plan` refuses to write a control plan that
misses the candidate's bytes by more than a control may (default 0.1%; the receipt's own
control is 65 ppm), and `verify` re-asserts on the two exported manifests' `wire_bytes`
before printing `BEAT` / `DID NOT BEAT`. On the receipt's numbers it prints `DID NOT BEAT`
at 1.996x. 27 tests; byte proof encode 0/14, decode 0/22.

Two design points that are decisions, not details: the control varies **only the rung**
(a BF16 passthrough is carried through untouched, and a two-family candidate refuses until
a grid is named), and the E2M1x2 axis has a **0.239 bpp hole** at the coset cap (R895 →
R896) where no byte-matched control exists — refused rather than papered over.

Carrying the verdict on the **shipcard** is the part left undone here; it is a
`shipcard.py` decision (advisory slot or blocking gate, and where the control's KL comes
from) and it should not arrive as a side effect of a Tessera batch. Filed with the exact
JSON block as #121.

### 3. "A cost that reproduces the served KL ordering of at least three Tessera rungs" — proposed, not built

Posted on RobTand/tessera#4: the four candidates ranked by cost to falsify, what falsifies
each, and two things worth naming here.

* **The cheapest discriminating experiment is a seed-variance screen**, not a re-price.
  The currency was measured at `nsamples: 4, max_act_rows: 256`; re-running the 21
  (unit, rung) `output_mse` measurements at four seeds bounds its own noise against the
  0.889 → 1.93 gap. If the spread is comparable, this is a resolution bug and that is the
  best available outcome; if it is small, candidate 4 is excluded before anything is spent
  on it.
* **The currency factorises, and the factorisation is testable offline.**
  `predicted_dloss = 0.5 · h_trace · weight_mse` (`allocator_solver.py:128`) with
  `h_trace` rung-independent, so within a unit the rung ordering is entirely
  `output_mse(r)` and between units the exchange rate is entirely `h_trace`. Once the
  served per-unit rung table exists, fitting the `h_q` that best reproduces it *given the
  measured curves* says which half is broken — and if a per-unit rescaling alone recovers
  the served ordering, the repair is the weights, not the objective.

**Acceptance, as this issue states it, is not met and was not claimed.** No cost model was
built, nothing was served, and the batch had no serving budget. What changed is that the
gate the acceptance criterion names is now callable and asserted, so the next candidate can
be judged instead of shipped.

**Comment by RobTand** [comment]
OpenCode triage (Muse Spark) — analysis only, not a maintainer verdict.

**Verdict: confirm (valid research issue; mostly cost-model work, partly tessera-side).** The served 2.00x-vs-uniform result is a measurement, not something re-derivable from this tree, so I verified the mechanism claims, not the KL numbers. The byte-accounting half is *not* verifiable on `main` @55d43e3.

**Evidence**
- `prismaquant/allocator_solver.py` exists; the batch comment cites `:128` (`predicted_dloss = 0.5 * h_trace * weight_mse`, `h_trace` rung-independent) — that factorisation is what makes the proposed offline test possible.
- `grep exact_bits_per_param|bits_for_shape_fn prismaquant/format_registry.py` → **no hits** on this checkout, and there is no `prismaquant/tessera_*.py` or `tests/test_tessera_footprint.py` here. So "FormatSpec fixed and pinned on main" describes the newer integration branch, not this `main`; confirm the overcharge fix on the branch actually merged before spending on cost work.
- The standing-gate half (byte-matched uniform arm beside every rate-axis result) is tessera-side (`tessera.control`, `experiments/uniform_control.py`) — wrong-repo for the build, right-repo for the gate; the shipcard half is #121.

**Proposed fix (prismaquant side)**
1. Cheapest screen first: re-run the 21 (unit, rung) `output_mse` measurements at 4 seeds; if spread ≈ the 0.889→1.93 gap, it's a resolution bug, not an objective bug.
2. If seed noise is small: fit per-unit `h_q` against the served per-unit rung table using already-measured curves — a rescaling that recovers served ordering indicts the weights, else the objective.
3. Acceptance: a cost reproducing served-KL ordering of ≥3 matched-byte Tessera rungs on one held-out model; every published rate-axis result ships with its byte-matched uniform arm.

**Comment by RobTand** [comment]
**Muse-Spark triage (automated first pass -- unverified)**

Branch read: `tessera/decouple-gridbook` at `0be02c4` (merge ahead of `main` at `90cd143`). Checked `git log --oneline -25` and `git branch -a` first — several fixes live only on this branch or on `main`-only ancestors, so "fixed on this branch" ≠ "fixed on `main`".

**Verdict: REAL, LOCATED — with one sub-claim STALE on this branch**

This issue bundles two mechanisms. They have opposite status on the branch triaged, so they must be separated:

**1. STALE on this branch (but REAL on `main`): the `FormatSpec` 6.25% overcharge.**
The issue body says `FormatSpec` special-cased Tessera and overcharged it by 6.25% on both the DP and the byte gate, and asks to confirm it is fixed before treating the ranking failure as a cost-model question. On the branch read it *is* fixed, on `main` it is not.

- *Measured fact, this branch:* `prismaquant/format_registry.py:112` declares `exact_bits_per_param: Fraction | None = None` with doc "is the WHOLE rate ... `memory_bytes_for_shape` uses it instead of, not in addition to" ; `prismaquant/format_registry.py:123` declares `bits_for_shape_fn` with doc "is the WHOLE size ... wins over `exact_bits_per_param`" ; `prismaquant/format_registry.py:250-270` implements that order — if `bits_for_shape_fn` set, `memory_bytes_for_shape` returns `ceil(bits/8)` from `bits_for_shape` (`:259-260`); else if `exact_bits_per_param` set returns `ceil(n_params*rate/8)` (`:261-267`) and never reaches `weight_bytes + scale_bytes` (`:268-270`). `prismaquant/format_registry.py:161-176` makes `effective_bits` refuse when `bits_for_shape_fn` is set. `prismaquant/format_registry.py:224-248` implements `bits_for_shape` as the single shape-aware accountant.
- *Measured fact, `main`:* `git show main:prismaquant/format_registry.py` has no `exact_bits_per_param` or `bits_for_shape_fn` fields at all (header-only, 90cd143), so the overcharge described in the issue still exists there.
- *Code-fact citation for fix:* commit `dfad8d1` ("Finish the merge the rename left half-done") on this branch history introduces the `exact_bits_per_param`/`bits_for_shape_fn` machinery and is an ancestor of `0be02c4` (`git log --oneline 0be02c4 | grep dfad8d1`). Tests `tests/test_tessera_footprint.py` and `tests/test_tessera_shape_dependent_recipe.py` exist on this branch and pin the accountant (issue comment cites 57 passed).
- *Inference:* the author's batch comment #5518947339 is correct that on this branch the DP was ranking correctly-priced things — the 2× served gap reproduced at 196/196 units with `1761722368 bits = 4.000260417 bpp` on charged vs wire bytes cannot be explained by byte accounting.

**2. REAL, LOCATED: the allocator's unary cost cannot rank rungs of one format at matched bytes — defect lives in `prismaquant/allocator_solver.py:128` and the caller that builds `Candidate.predicted_dloss`.**

- *Code fact, defect site:* `prismaquant/allocator_solver.py:128-131` `def predicted_dloss(h_trace: float, weight_mse: float, gain: float = 1.0) -> float: return 0.5 * float(h_trace) * float(weight_mse) * float(gain)`. `h_trace` is per-unit, rung-independent. The batch comment correctly factorises it (`allocator_solver.py:128` comment cited) as `predicted_dloss = 0.5 · h_trace · weight_mse` with `h_trace` constant within a unit — therefore within a unit the rung ordering is entirely `weight_mse`/`output_mse`, between units the exchange rate is entirely `h_trace`. No term sees rate-allocation or cross-unit interaction.
- *Code fact, why this fails at matched bytes:* `prismaquant/allocator_candidates.py` builds each `Candidate.predicted_dloss` from that unary product (via `cost_entry_predicted_dloss` at `:1274-1318` delegating to `predicted_dloss`). The DP in `allocator_solver.py:639-748` `solve_allocation` maximizes `sum dgain` under `sum dbins` with `dgain = baseline_loss - c.predicted_dloss`. At matched bytes the Tessera rungs differ by small, smooth deltas dominated by allocation effects the unary cost does not measure — exactly the issue body diagnosis ("AURA's `predicted_dloss` is calibrated across *formats* with very different error structure").
- *Measured fact, served gap is not re-derivable from this tree alone:* issue reports a served checkpoint at 4.0 bpp 2.00× worse in KL than a byte-matched uniform arm, with byte-exact allocation loop (196/196 bytes reproduce, re-encode byte-identical). No code here can reproduce that KL number without the serving trace; I verified the *mechanism* claim (cost factorisation) not the KL magnitude. The mechanism reproduces in code as the unary factorization above.
- *Author's own batch disposition confirms acceptance not met:* comment #5518947339 section 3 states "A cost that reproduces the served KL ordering of at least three Tessera rungs — proposed, not built" and explicitly "Acceptance, as this issue states it, is not met and was not claimed." So this is not STALE.
- *Prior automated triage partially duplicates but under-states branch divergence:* the OpenCode triage comment correctly noted the factorisation and the byte-gate being tessera-side, but it claimed the byte-accounting half is "not verifiable on `main` @55d43e3" without distinguishing that it *is* verifiable (and fixed) on the `tessera/decouple-gridbook` branch actually checked out — the distinction the task description warns about.

**Mechanism in one paragraph:** At `prismaquant/allocator_solver.py:128` the DP's objective is the diagonal-Fisher product `0.5·h_trace·weight_mse` with `h_trace` constant per unit. Within a fused unit every Tessera rung of the same family shares that constant, so intra-family ranking collapses to ordering by `weight_mse`/`output_mse`. At matched wire bytes the true served KL ordering is driven by rate-allocation smoothness, cross-column interaction, and activation-path differences the weight-only (or output-mse-on-small-calibration) error does not resolve. At the measurement scale cited (`nsamples: 4, max_act_rows: 256`) the surrogate's resolution is insufficient to separate rungs 0.889→1.93 apart, so the DP spends budget on rungs with lower unary cost but higher served KL, producing the served 1.996× DID NOT BEAT vs the 65 ppm byte-matched uniform control.

**Proposed fix (description, not a diff):**

- *Function to change:* `prismaquant/allocator_candidates.py` `cost_entry_predicted_dloss`/`build_candidates` (which price `Candidate.predicted_dloss`) and/or a new cost estimator alongside `prismaquant/allocator_solver.py:128` `predicted_dloss`. Do not change `prismaquant/format_registry.py` — byte accounting is already correct on this branch.
- *New behaviour:* Introduce a cost that reproduces served-KL *ordering* of ≥3 Tessera rungs at matched bytes on one held-out model (issue acceptance). The cheapest falsifiable path per the batch — already posted to `RobTand/tessera#4` — is: (a) seed-variance screen: re-run the 21 `(unit, rung)` `output_mse` measurements at 4 seeds; if spread ≈ the 0.889→1.93 gap, the surrogate is resolution-limited, not structurally wrong; (b) if seed noise is small, fit per-unit `h_q` against the served per-unit rung table using the already-measured curves — `predicted_dloss = 0.5·h_q·output_mse(r)` — and test whether a per-unit rescaling alone recovers the served ordering (weights fix) vs requiring a new objective; (c) acceptance gate per `RobTand/tessera#3`: every rate-axis candidate is measured beside a byte-matched uniform arm via `tessera.control` + `experiments/uniform_control.py` (`plan` refuses >0.1% byte mismatch, `verify` re-asserts `wire_bytes` and prints `BEAT`/`DID NOT BEAT`), carried as an advisory/blocking field on the shipcard per #121. Until the surrogate passes that gate, selection runs as `SELECTION_MODE=validated-surrogate` (real held-out KL), not surrogate-selected — consequence already stated in issue body.
- *Test that would fail before and pass after:* A new test (tessera-side or `tests/test_tessera_*.py`) that loads three byte-matched Tessera rungs from one allocation (wire bytes within 0.1% via `tessera.control.unit_wire_bits`), obtains their served per-unit KL order from a held-out run, and asserts the *new* cost's ordering matches served ordering; today with `0.5·h_trace·weight_mse` it fails (inverts or ties the middle rung), after the per-unit `h_q` or revised objective it passes. A second test asserts every published rate-axis result ships with its byte-matched uniform arm and that the uniform arm is not beaten (the gate from tessera#3), which today has no producer-side enforcement.

### Confidence
medium — Byte-accounting fix verified line-by-line on this branch (`prismaquant/format_registry.py:112`, `:123`, `:250-270`) and its absence on `main` (90cd143) confirmed via `git show`. Cost factorisation and DP objective verified at `prismaquant/allocator_solver.py:128` and `prismaquant/allocator_candidates.py:1274-1350`. I did not re-serve the 2.00× KL number — it is a measurement claim and would require the served artifact + the byte-matched control plan from `RobTand/tessera#3`'s unpushed worktree branch, which is not in this tree; I treated it as reported. High confidence the issue is REAL (not STALE) on this branch for the cost-model half; high confidence the byte-accounting sub-issue is STALE on this branch but still REAL on `main`.

**Comment by RobTand** [comment]
Landed. The closure is a gate at the selection point, not a cost change: refused by default, acknowledged per run with a stamped run id (so the candidate the control is built FROM can be built), not-applicable when the plan is itself uniform. Publication still refuses until the served control closes the shipcard slot (#121). The `measurement-needed` half — the served byte-matched A/B — belongs to tessera#1 and stays open there.
